### ISSUE TITLE
Train in half precission

### ISSUE THREAD

**Issue by hvico** [EricFillion/happy-transformer#257]
Hello,

It would be great to have a training mode or arg which would enable FP16 mode, for enhanced performance on CUDA traininig.

I use this training code for regular hugginface training:

`scaler = torch.cuda.amp.GradScaler()

def train_fp16(epoch):
    model.train()
    for _,data in enumerate(training_loader, 0):
        ids = data['ids'].to(device, dtype = torch.long)
        mask = data['mask'].to(device, dtype = torch.long)
        token_type_ids = data['token_type_ids'].to(device, dtype = torch.long)
        targets = data['targets'].to(device, dtype = torch.float)

        with torch.cuda.amp.autocast():

            outputs = model(ids, mask, token_type_ids)

        optimizer.zero_grad()
        loss = loss_fn(outputs, targets)
        #if _%1000==0:
        #    print(f'Epoch: {epoch}, Loss:  {loss.item()}')
        
        optimizer.zero_grad()
        scaler.scale(loss).backward()
        scaler.step(optimizer)
        scaler.update()
        
def train_fp32(epoch):
    model.train()
    for _,data in enumerate(training_loader, 0):
        ids = data['ids'].to(device, dtype = torch.long)
        mask = data['mask'].to(device, dtype = torch.long)
        token_type_ids = data['token_type_ids'].to(device, dtype = torch.long)
        targets = data['targets'].to(device, dtype = torch.float)

        outputs = model(ids, mask, token_type_ids)

        optimizer.zero_grad()
        loss = loss_fn(outputs, targets)
        if _%1000==0:
            print(f'Epoch: {epoch}, Loss:  {loss.item()}')
        
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()`

Thanks!

**Comment by EricFillion** [comment]
Thanks for the suggestion! I'll look into it. I agree with adding more features to reduce memory consumption.

**Comment by EricFillion** [comment]
Once again, thanks for this suggestion. I just published a [new version ](https://pypi.org/project/happytransformer/2.4.0/)that allows you to enable fp16.
